### PR TITLE
Spidevwrite

### PR DESCRIPTION
**Pull request by rminnich**
Believe it or not, this code worked last week. And not today. But Stefan and I did a bit of soldering action on the QSPI MUX and, who knows, maybe we busted it. But the logic analyzer shows the right things happening. 

I think this is a forward step, but if you see an error, please let me know.

### CODE REVIEW

**Comment by rminnich** [comment]
This is a mystery, as to why it worked last week and no longer. But in between then and now, we did do some board rework so ... we may have damaged something. 
Here's the trace for writing a byte
Screenshot 2024-01-16 at 10 36 45 PM

**Comment by rminnich** [comment]
With this change, write looks like this, but still does not work
Screen Shot 2020-10-06 at 2 52 55 PM

**Comment by rminnich** [comment]
I also just verified that the qemu mux is good, by hooking a dediprog sf100 to it. 

I did not need to hook up hold and wp, btw, so this looks purely like an error on my part.

File: pkg/flash/flash_linux.go [hunk 1]
```diff
@@ -104,7 +106,7 @@ func (f *Flash) Size() int64 {
 
 const maxTransferSize = 4096
 
-func min(x, y int64) int64 {
```
Comment (rjoleary) [inline pkg/flash/flash_linux.go:107]: For some reason, `min` is now a language built-in since go 1.21. https://go.dev/play/p/rZtnXH0hPtN

Comment (rjoleary) [reply pkg/flash/flash_linux.go:107]: Really, I had to do a double-take. They added min/max functions to the language itself (https://go.dev/ref/spec#Min_and_max) and not the just standard library.

File: pkg/flash/flash_linux.go [hunk 2]
```diff
@@ -149,10 +151,19 @@ func (f *Flash) writeAt(p []byte, off int64) (int, error) {
...
 		// Send the address.
-		{Tx: append(op.PageProgram.Bytes(), f.prepareAddress(off)...)},
-		// Send the data.
-		{Tx: p, CSChange: true},
-		{Tx: op.WriteDisable.Bytes(), CSChange: true},
+		{Tx: append(append(op.PageProgram.Bytes(), f.prepareAddress(off)...), p...)},
+	}); err != nil {
+		return 0, err
+	}
+	time.Sleep(10 * time.Microsecond)
```
Comment (rjoleary) [inline pkg/flash/flash_linux.go:158]: The `Transfer` struct has a `DelayUSecs` option. `DelayUSecs is how long to delay before the next transfer (optional).`. Then you won't be affected by userspace scheduling.

File: pkg/flash/flash_linux.go [hunk 3]
```diff
@@ -149,10 +151,19 @@ func (f *Flash) writeAt(p []byte, off int64) (int, error) {
...
+	if err := f.spi.Transfer([]spidev.Transfer{
+		// The meaning of CSChange is ... odd.
+		// IF CSChange is set true here, then CE# never goes
```
Comment (rjoleary) [inline pkg/flash/flash_linux.go:161]: I think cschange has a different meaning on the last transaction. Let me look at the kernel source and get back...

Comment (rjoleary) [reply pkg/flash/flash_linux.go:161]: If you `grep cs_change drivers/spi/spi.c`, you'll see the kernel does some extra stuff when it sees this bit, such as adding additional delays.

**Comment by rminnich** [comment]
OK, turns out it works! 
The BPL bits were set. In a forthcoming PR I will add code to manage them. But I just wrote the flash by doing this.
echo -n '\x0100' | spidev raw
that clears SR, and clears block protects.
Then:
flash -p linux_spi:dev=/dev/spidev0.0   -w /etc/hosts
flash -p linux_spi:dev=/dev/spidev0.0   -r xyz
hexdump -C xyz | less
/etc/hosts is visible

File: pkg/flash/flash_linux.go (hunk 2 above)
Comment (rminnich) [reply pkg/flash/flash_linux.go:158]: I think I'm going to replace this by polling the SR.

File: pkg/flash/flash_linux.go (hunk 3 above)
Comment (rminnich) [reply pkg/flash/flash_linux.go:161]: confusing. But at least this works ...

File: pkg/flash/flash_linux.go (hunk 1 above)
Comment (rminnich) [reply pkg/flash/flash_linux.go:107]: finally.

Comment (rminnich) [reply pkg/flash/flash_linux.go:107]: I bet they added it because everybody write them

Comment (rminnich) [reply pkg/flash/flash_linux.go:107]: Anyway, we have to leave this here until the CI runs on 1.21 uniformly. We're stuck on 1.20 on a few targets still.

File: pkg/flash/flash_linux.go (hunk 2 above)
Comment (rminnich) [reply pkg/flash/flash_linux.go:158]: Never mind, you are right. Delay Usecs is a good way to go.

For now, we can write the data at one byte per (say) 10 microseconds + 8 or so bytes of transfer, which is not horrendous. We can make it faster later.

File: cmds/fwtools/flash/flash_linux.go [hunk 4]
```diff
@@ -155,21 +155,19 @@ func run(args []string, supportedProgrammers map[string]programmerInit) (reterr
...
+		buf, err := os.ReadFile(*w)
 		if err != nil {
 			return err
 		}
-		defer f.Close()
-		if _, err := io.ReadFull(f, buf); err != nil {
-			return err
+		amt, err := programmer.WriteAt(buf, 0)
+		if err != nil {
+			return fmt.Errorf("Writing %d bytes to dev %v:%w", len(buf), programmer, err)
```
Comment (binjip978) [inline cmds/fwtools/flash/flash_linux.go:164]: Error strings should not be capitalized (ST1005).

File: cmds/fwtools/flash/flash_linux.go [hunk 5]
```diff
@@ -155,21 +155,19 @@ func run(args []string, supportedProgrammers map[string]programmerInit) (reterr
...
 		}
-		if leftover, err := io.Copy(io.Discard, f); err != nil {
-			return err
-		} else if leftover != 0 {
-			return fmt.Errorf("flash size (%#x) unequal to file size (%#x)", len(buf), int64(len(buf))+leftover)
+		if amt != len(buf) {
+			return fmt.Errorf("Only flashed %d of %d bytes", amt, len(buf))
```
Comment (binjip978) [inline cmds/fwtools/flash/flash_linux.go:167]: Error strings should not be capitalized (ST1005).

**Review by binjip978 — APPROVED** [review]

**Pull request merged**